components/eam/src/physics/cam/micro_p3.F90

```diff
@@ -2855,6 +2868,125 @@ subroutine prevent_ice_overdepletion(pres,t_atm,qv,latent_heat_sublim,inv_dt,
+  real(rtype)                :: qv_avail, qv_sink, ratio
+
+  qv_avail = qv + (qi2qv_sublim_tend+qr2qv_evap_tend)*dt
+  qv_sink  = (qidep + qinuc)*dt
+
+  if (qv_sink > qv_avail .and. qv_sink>1.e-20_rtype) then
+     ratio = qv_avail/qv_sink
+     qidep = qidep*ratio
+     qinuc = qinuc*ratio     
+  endif
+
+  return
+end subroutine water_vapor_conservation
+
+subroutine ice_supersat_conservation(qidep,qinuc,cld_frac_i,qv,qv_sat_i,latent_heat_sublim,T_atm,dt)
+  !Make sure ice processes don't drag qv below ice supersaturation
+  
+  implicit none
+
+  real(rtype), intent(in) :: cld_frac_i,qv,qv_sat_i,latent_heat_sublim,T_atm,dt
+  real(rtype), intent(inout) :: qidep,qinuc
+  
+  real(rtype) :: qv_sink, qv_avail, fract 
+
+  qv_sink = qidep + qinuc ! in [kg/kg] cell-avg values
+
+  if (qv_sink > qsmall .and. cld_frac_i > 1.0e-20_rtype) then
+     ! --- Available water vapor for deposition/nucleation
+     qv_avail = (qv - qv_sat_i) / &
+          (1.0_rtype + latent_heat_sublim**2.0_rtype*qv_sat_i / (cp*rv*T_atm**2.0_rtype) ) / dt
+
+     ! --- Only excess water vapor can be limited
+     qv_avail = max(qv_avail,0.0_rtype)
+         
+     if (qv_sink >  qv_avail) then
+        fract = qv_avail / qv_sink
+        qinuc = qinuc * fract
+        qidep = qidep * fract
+     endif
+  endif
+
+  return
+end subroutine ice_supersat_conservation
+
+subroutine nc_conservation(nc, nc_selfcollect_tend, dt, nc_collect_tend, nc2ni_immers_freeze_tend, &
+     nc_accret_tend, nc2nr_autoconv_tend)
+  !Make sure nc doesn't go below zero
+  
+  implicit none
+  
+  real(rtype), intent(in) :: nc,nc_selfcollect_tend,dt
+  real(rtype), intent(inout) :: nc_collect_tend,nc2ni_immers_freeze_tend,&
+                                nc_accret_tend,nc2nr_autoconv_tend
+  real(rtype) :: sink_nc, source_nc, ratio
+  
+  sink_nc = (nc_collect_tend + nc2ni_immers_freeze_tend + nc_accret_tend + nc2nr_autoconv_tend)*dt
+  source_nc = nc + nc_selfcollect_tend*dt
+  if(sink_nc > source_nc) then
+     ratio = source_nc/sink_nc
+     nc_collect_tend  = nc_collect_tend*ratio
+     nc2ni_immers_freeze_tend = nc2ni_immers_freeze_tend*ratio 
+     nc_accret_tend  = nc_accret_tend*ratio 
+     nc2nr_autoconv_tend = nc2nr_autoconv_tend*ratio
+  endif
+  
+  return
+end subroutine nc_conservation
+
+subroutine nr_conservation(nr,ni2nr_melt_tend,nr_ice_shed_tend,ncshdc,nc2nr_autoconv_tend,dt,nr_collect_tend,&
+     nr2ni_immers_freeze_tend,nr_selfcollect_tend,nr_evap_tend)
+  !Make sure nr doesn't go below zero
+
+  implicit none
+
+  real(rtype), intent(in) :: nr,ni2nr_melt_tend,nr_ice_shed_tend,ncshdc,nc2nr_autoconv_tend,dt
+  real(rtype), intent(inout) :: nr_collect_tend,nr2ni_immers_freeze_tend,nr_selfcollect_tend,nr_evap_tend
+  real(rtype) :: sink_nr, source_nr, ratio
+
+  sink_nr = (nr_collect_tend + nr2ni_immers_freeze_tend + nr_selfcollect_tend + nr_evap_tend)*dt 
+  source_nr = nr + (ni2nr_melt_tend + nr_ice_shed_tend + ncshdc + nc2nr_autoconv_tend)*dt
```

Comment:
nc2nr_autoconv_tend should be ncautr here as well. 

nr_collect_tend and nr2ni_immers_freeze_tend, ni2nr_melt_tend, nr_ice_shed_tend, ncshdc all don't appear in update_prognostic_liquid so either should be added there, removed here, or I should figure out what's going on.

Reply:
I agree, nc2nr_autoconv_tend should be ncautr here. 

The terms you listed are all ice processes, I think, and are oddly listed under update_prognostic_ice (perhaps the phase connotes liquid only processes vs any process including ice). So that's why they are missing there. I think it's okay as it is.

Reply:
Ok, switched to ncautr here. You're right about the quantities I'm worried about just showing up in the update_prognostic_ice instead of update_prognostic_liquid. I think it is kind of ambiguous whether a process taking liquid and converting it to ice should be in update_prognostic_liquid or ice... I think it is fine as is.